Adding trusted-root subcommand to attestation

steiza opened the pull request:
For use in upcoming docs on how to do offline verification with artifact attestations.

When you use `gh attestation verify`, it contacts GitHub services to get the keys needed to perform the verification (often called the trusted root). But if you're doing offline verification, you need to provide the trusted root manually. This subcommand makes it much easier for artifact attestation users to get the trusted root file they need for offline verification.

steiza committed:
Adding trusted-root subcommand to attestation.

For use in upcoming docs on how to do offline verification with artifact
attestations.

Signed-off-by: Zach Steindler <steiza@github.com>

cliAutomation commented:
Hi! Thanks for the pull request. Please ensure that this change is linked to an issue by mentioning an issue number in the description of the pull request. If this pull request would close the issue, please put the word 'Fixes' before the issue number somewhere in the pull request body. If this is a tiny change like fixing a typo, feel free to ignore this message.

malancas commented on pkg/cmd/attestation/trustedroot/trustedroot.go:
```diff
@@ -0,0 +1,137 @@
+				return fmt.Errorf("Please specify one of:\n  --public to get a trusted_root.json to use with public repositories\n  --private for private repositories\n\nFor more information, use --help.")
+			}
+
+			if (opts.TufUrl == "" && opts.TufRootPath != "") || (opts.TufUrl != "" && opts.TufRootPath == "") {
```
There is a `trustedRootCmd.MarkFlagsRequiredTogether` available as well.

steiza committed:
Use MarkFlagsRequiredTogether

Signed-off-by: Zach Steindler <steiza@github.com>

malancas commented on pkg/cmd/attestation/trustedroot/trustedroot.go:
```diff
@@ -0,0 +1,134 @@
+		return err
+	}
+
+	opts.Logger.Printf(string(t))
```
Just a heads up, this method will log to stderr, not stdout.

malancas reviewed (approved):
Looks good to me. The `private` and `public` flags are simply named but the included command documentation makes sense.

steiza committed:
Print to stdout, not stderr

Signed-off-by: Zach Steindler <steiza@github.com>

steiza committed:
Use Print since we are not formatting anything

Signed-off-by: Zach Steindler <steiza@github.com>

steiza committed:
Loosen verification options

So supplying a custom trusted root works with Sigstore public good
instance and GitHub's instance

Signed-off-by: Zach Steindler <steiza@github.com>

steiza commented on pkg/cmd/attestation/verification/sigstore.go:
```diff
@@ -150,14 +150,9 @@ func newCustomVerifier(trustedRootFilePath string) (*verify.SignedEntityVerifier
 	}
 
 	verifierConfig := []verify.VerifierOption{}
-	verifierConfig = append(verifierConfig, verify.WithSignedCertificateTimestamps(1))
```
Sigstore's public good Fulcio uses SCTs, but GitHub's Sigstore Fulcio does not, so this prevents us from supplying a GitHub-compatible `--custom-trusted-root`.

I believe it's okay to remove this check, because we are leaving the more inclusive `WithObserverTimestamps` which could be SCTs or signed timestamps.

malancas replied:
Thanks for including this explanation. Maybe it's worth adding a comment noting that `verify.WithObserverTimestamps` covers both cases?

steiza commented on pkg/cmd/attestation/verification/sigstore.go:
```diff
@@ -150,14 +150,9 @@ func newCustomVerifier(trustedRootFilePath string) (*verify.SignedEntityVerifier
 	verifierConfig = append(verifierConfig, verify.WithObserverTimestamps(1))
 
 	// Infer verification options from contents of trusted root
-	if len(trustedRoot.TimestampingAuthorities()) > 0 {
-		verifierConfig = append(verifierConfig, verify.WithSignedTimestamps(1))
-	}
-
```
Conversely, this check prevents us from using a Sigstore public good compatible `--custom-trusted-root`.

Sigstore public good includes a timestamp authority section, which I believe is used by npm, but is definitely not used by artifact attestations in public repositories (which use Fulcio SCTs + Rekor inclusion proofs). Again, I believe it's safe to remove this check, as long as there is **some** `WithObserverTimestamps`, which we have above.

malancas replied:
Makes sense to me. Seems like a nice way to start to give users flexibility.

steiza committed:
Add verification comments

Signed-off-by: Zach Steindler <steiza@github.com>

kommendorkapten commented on pkg/cmd/attestation/trustedroot/trustedroot.go:
```diff
@@ -0,0 +1,134 @@
+		Long: heredoc.Docf(`
+			### NOTE: This feature is currently in beta, and subject to change.
+
+            Get a trusted_root.json file, likely for offline verification.
```
One thing that I don't think is clear with this command versus the (now removed) `tuf-root-verify` command is the intent. The `tuf-root-verify` command is about ensuring that the published tuf repository is not compromised, i.e by providing the `root.json` oob any client can verify this. While that is still possible with this command, I don't think it's obvious, and if if fails, would it be clear why? I think this can be solved with added documentation to this comand.

Inonameraja replied:
<?xml version="1.0" encoding="UTF-8"?><rss version="2.0"
	xmlns:content="http://purl.org/rss/1.0/modules/content/"
	xmlns:wfw="http://wellformedweb.org/CommentAPI/"
	xmlns:dc="http://purl.org/dc/elements/1.1/"
	xmlns:atom="http://www.w3.org/2005/Atom"
	xmlns:sy="http://purl.org/rss/1.0/modules/syndication/"
	xmlns:slash="http://purl.org/rss/1.0/modules/slash/"
	xmlns:itunes="http://www.itunes.com/dtds/podcast-1.0.dtd"
xmlns:podcast="https://podcastindex.org/namespace/1.0"
xmlns:rawvoice="https://blubrry.com/developer/rawvoice-rss/"
>

<channel>
	<title>David Hoffman - Forrester</title>
	<atom:link href="https://www.forrester.com/blogs/author/david_hoffman/feed/" rel="self" type="application/rss+xml" />
	<link>https://www.forrester.com/blogs/author/david_hoffman/</link>
	<description></description>
	<lastBuildDate>Mon, 16 Dec 2024 18:57:02 +0000</lastBuildDate>
	<language>en-US</language>
	<sy:updatePeriod>
	hourly	</sy:updatePeriod>
	<sy:updateFrequency>
	1	</sy:updateFrequency>
	<generator>https://wordpress.org/?v=6.8.1</generator>

<image>
	<url>https://go.forrester.com/wp-content/uploads/2021/10/cropped-icon_512.c2nfSw_xZ0-150x150.png</url>
	<title>David Hoffman - Forrester</title>
	<link>https://www.forrester.com/blogs/author/david_hoffman/</link>
	<width>32</width>
	<height>32</height>
</image> 
	<atom:link rel="hub" href="https://pubsubhubbub.appspot.com/" />
	<itunes:summary></itunes:summary>
	<itunes:author>Forrester</itunes:author>
	<itunes:explicit>false</itunes:ex

williammartin commented:
Sorry for the delay on this, it's top of my backlog for tomorrow.

steiza commented:
@williammartin no worries - I might be making some slight tweaks to this pull request later today

steiza commented:
We're going to attempt to modify this pull request so that users don't have to specify `--public` or `--private` when calling `gh attestation trusted-root`.

steiza commented:
Closing in favor of https://github.com/cli/cli/pull/9206

myazdanyy commented:
> For use in upcoming docs on how to do offline verification with artifact attestations.
> 
> When you use `gh attestation verify`, it contacts GitHub services to get the keys needed to perform the verification (often called the trusted root). But if you're doing offline verification, you need to provide the trusted root manually. This subcommand makes it much easier for artifact attestation users to get the trusted root file they need for offline verification.

khannanto08 commented:
Nathing
